### PR TITLE
Added the train loop

### PR DESCRIPTION
**Pull request by sayannath**
- [x] Build the SAGAN module
- [x] Define the optimizer and loss function
- [x] Code the train loop

### CODE REVIEW

File: sagan/train.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+import matplotlib.pyplot as plt
+
+import tensorflow as tf
+from tensorflow import keras
+
+from models.generator import build_generator
+from models.discriminator import build_discriminator
+
+
+## Define the hyperparameters
```
Comment (soumik12345) [inline sagan/train.py]: Hyperparameters should be defined separately in a config file. Refer to https://github.com/soumik12345/point-cloud-segmentation/blob/main/configs/shapenetcore.py

File: sagan/train.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+BATCH_SIZE = 64
+LATENT_DIM = 128
+EPOCHS = 100
+generator_save_path = "./saved_models/generator.h5"
+discriminator_save_path = "./saved_models/discriminator.h5"
+
+## Save the images
```
Comment (soumik12345) [inline sagan/train.py]: utilities should ideally be in a separate file such as `utils.py`. Refer to https://github.com/soumik12345/point-cloud-segmentation/blob/main/point_seg/utils.py

File: sagan/train.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        plt.axis("off")
+        plt.imshow(examples[i])
+    filename = f"/samples/demo-{epoch+1}.png"
+    plt.savefig(filename)
+    plt.close()
+
+
+# Build the SAGAN module
+class SAGAN(keras.Model):
+    def __init__(self, discriminator, generator, latent_dim, **kwargs):
```
Comment (soumik12345) [inline sagan/train.py]: The config data structure should be taken as input.

**Review by soumik12345 — CHANGES_REQUESTED** [review]
@sayannath Gradient Penalty seems to be missing

**Comment by sayannath** [comment]
@soumik12345 PTAL!

File: configs/config.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import ml_collections
+
+
+def get_config() -> ml_collections.ConfigDicts:
+    config = ml_collections.ConfigDict()
+    config.batch_size = 64  # Batch Size
+    config.epochs = 100  # Number of Epochs
+    config.image_size = 128  # Image Size
+
+    config.generator_learning_rate = 0.0001  # Generator Learning Rate
```
Comment (soumik12345) [inline configs/config.py]: Instead of writing `0.0001` write `1e-4`, makes it more readable.

File: configs/config.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    config.discriminator_learning_rate = 0.0004  # Discriminator Learning Rate
```
Comment (soumik12345) [inline configs/config.py]: Instead of writing `0.0001` write `1e-4`, makes it more readable.

File: configs/config.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
+import ml_collections
```
Comment (soumik12345) [inline configs/config.py]: instead of naming the config file as `config.py`, name it something like `base.py`, there could be a couple of different config files.

File: sagan/train.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+LATENT_DIM = 128
+EPOCHS = 100
+generator_save_path = "./saved_models/generator.h5"
+discriminator_save_path = "./saved_models/discriminator.h5"
+
+# Build the SAGAN module
+class SAGAN(keras.Model):
+    def __init__(self, discriminator, generator, latent_dim, gp_weight=10.0, **kwargs):
+        super(SAGAN, self).__init__(**kwargs)
+        self.discriminator = discriminator
```
Comment (soumik12345) [inline sagan/train.py]: We should initialize the discriminator in the `SAGAN` model instead of passing it as a parameter.

File: sagan/train.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        self.generator = generator
```
Comment (soumik12345) [inline sagan/train.py]: We should initialize the generator in the `SAGAN` model instead of passing it as a parameter.

File: sagan/train.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+    def compile(self, d_optimizer, g_optimizer, loss_fn):
+        super(SAGAN, self).compile()
+        self.d_optimizer = d_optimizer
+        self.g_optimizer = g_optimizer
+        self.loss_fn = loss_fn
+
+    # Calculate the gradient_penalty.
+    def gradient_penalty(self, batch_size, real_images, fake_images):
+        # Get the interpolated image
```
Comment (soumik12345) [inline sagan/train.py]: Add a reference to https://keras.io/examples/generative/wgan_gp/#create-the-wgangp-model

File: sagan/train.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+        grads = gp_tape.gradient(pred, [interpolated])[0]
+        # 3. Calculate the norm of the gradients.
+        norm = tf.sqrt(tf.reduce_sum(tf.square(grads), axis=[1, 2, 3]))
+        gp = tf.reduce_mean((norm - 1.0) ** 2)
+        return gp
+
+    def train_step(self, real_images):
+        batch_size = tf.shape(real_images)[0]
+
+        for _ in range(2):
```
Comment (soumik12345) [inline sagan/train.py]: why 2?

File: sagan/train.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+            # Sample random points in the latent space
+            random_latent_vectors = tf.random.normal(
```
Comment (soumik12345) [inline sagan/train.py]: I think we should make it `tf.random.truncated_normal`

File: sagan/train.py [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+                shape=(batch_size, self.latent_dim)
+            )
+            generated_images = self.generator(random_latent_vectors)
```
Comment (soumik12345) [inline sagan/train.py]: Note the generator model also returns the attentions

```python3
generated_images, generator_attention_1, generator_attention_2 = self.generator(random_latent_vectors, training=False)
```

File: sagan/train.py [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+            generated_labels = tf.zeros((batch_size, 1))
+
+            # Train the discriminator
+            with tf.GradientTape() as ftape:
+                predictions = self.discriminator(generated_images)
```
Comment (soumik12345) [inline sagan/train.py]:
```python3
fake_prediction, fake_attention_1, fake_attention_2 = self.discriminator(generated_images, training=True)
```

File: sagan/train.py [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+                d1_loss = self.loss_fn(generated_labels, predictions)
+            grads = ftape.gradient(d1_loss, self.discriminator.trainable_weights)
+            self.d_optimizer.apply_gradients(
+                zip(grads, self.discriminator.trainable_weights)
+            )
+
+            labels = tf.ones((batch_size, 1))
+
+            with tf.GradientTape() as rtape:
+                predictions = self.discriminator(real_images)
```
Comment (soumik12345) [inline sagan/train.py]:
```python3
real_prediction, real_attention_1, real_attention_2 = self.discriminator(generated_images, training=True)
```

File: sagan/train.py [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+            self.d_optimizer.apply_gradients(
+                zip(grads, self.discriminator.trainable_weights)
+            )
+
+        ## Train the generator
+        random_latent_vectors = tf.random.normal(shape=(batch_size, self.latent_dim))
+        misleading_labels = tf.ones((batch_size, 1))
+
+        with tf.GradientTape() as gtape:
+            predictions = self.discriminator(self.generator(random_latent_vectors))
```
Comment (soumik12345) [inline sagan/train.py]:
```python3
generated_images, g_attention_1, g_attention_2  = self.generator(random_latent_vectors, training=True)
fake_prediction, d_attention_1, d_attention_2 = self.discriminator(generated_images, training=False)
```

File: sagan/train.py [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+            g_loss = self.loss_fn(misleading_labels, predictions)
+        grads = gtape.gradient(g_loss, self.generator.trainable_weights)
+        self.g_optimizer.apply_gradients(zip(grads, self.generator.trainable_weights))
+
+        return {"d1_loss": d1_loss, "d2_loss": d2_loss, "g_loss": g_loss}
```
Comment (soumik12345) [inline sagan/train.py]: Return total discriminator loss instead of separtely. I think we should also return the gradient penalty here.

File: sagan/train.py [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+)
+
+### Define the loss function
+loss_fn = tf.keras.losses.BinaryCrossentropy(from_logits=True, label_smoothing=0.1)
+
+sagan = sagan.compile(discriminator_optimizer, generator_optimizer, loss_fn)
+
+## Train the model
+for epoch in range(EPOCHS):
+    sagan.fit(dataset, epochs=25)
```
Comment (soumik12345) [inline sagan/train.py]: ???

**Review by soumik12345 — CHANGES_REQUESTED** [review]
@sayannath Please take a look at https://github.com/soumik12345/self-attention-gan/issues/5 as well.

**Comment by soumik12345** [comment]
@sayannath I think the workflow is failing because of the caltech birds dataset. Try changing it to `tf_flowers` for the time and check if the workflow passes or not.

**Comment by soumik12345** [comment]
@sayannath A few more suggestions:

- let's move the SAGAN model inside the `models` module and rename it to `SelfAttentionGAN`
- we will create a separate CLI driven script for training separate from the module like https://github.com/soumik12345/point-cloud-segmentation/blob/main/train_shapenet_core.py

File: sagan/models/sagan.py [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+from models.discriminator import build_discriminator
+from sagan.utils import w_loss
+
+class SelfAttentionGAN(keras.Model):
+
+    def __init__(self, latent_dim, gp_weight, **kwargs):
+        super(SelfAttentionGAN, self).__init__(**kwargs)
+        self.latent_dim = latent_dim
+        self.gp_weight = gp_weight
+        self.generator = build_generator(IMAGE_SIZE, latent_dim)
```
Comment (soumik12345) [inline sagan/models/sagan.py]: As mentioned in https://github.com/soumik12345/self-attention-gan/pull/3#discussion_r744122023, the config object should be passed to the constructor instead of using constants. Refer to https://github.com/soumik12345/point-cloud-segmentation

File: sagan/models/sagan.py [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+        self.discriminator = build_discriminator(IMAGE_SIZE)
+    
+    def compile(self, g_optimizer, d_optimizer, **kwargs):
+        super(SelfAttentionGAN, self).compile(**kwargs)
+        self.g_optimizer = g_optimizer
+        self.d_optimizer = d_optimizer
+    
+    # Reference from https://keras.io/examples/generative/wgan_gp/#create-the-wgangp-model
+    def gradient_penalty(self, batch_size, real_images, fake_images):
```
Comment (soumik12345) [inline sagan/models/sagan.py:29]: Reference should be added in the form of a docstring, not a generic command.

```python3
def gradient_penalty(self, batch_size, real_images, fake_images):
    """Reference: https://keras.io/examples/generative/wgan_gp/#create-the-wgangp-model"""
```

File: sagan/train.py [hunk 20]
```diff
@@ -0,0 +1,46 @@
+import numpy as np
```
Comment (soumik12345) [inline sagan/train.py]: `train.py` should be outside the `sagan` module

**Review by soumik12345 — CHANGES_REQUESTED** [review]
@sayannath Feedback round 2

File: sagan/train.py (hunk 20 above)
Comment (sayannath) [reply sagan/train.py]: Cool

**Comment by sayannath** [comment]
@soumik12345 PTAL! Just got one issue

File: sagan/models/sagan.py [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+
+from models.generator import build_generator
+from models.discriminator import build_discriminator
+from sagan.utils import w_loss
+
+FLAGS = flags.FLAGS
+config_flags.DEFINE_config_file("experiment_configs")
+
+class SelfAttentionGAN(keras.Model):
+    def __init__(self, FLAGS.experiment_configs.latent_dim, FLAGS.experiment_configs.gp_weight, **kwargs):
```
Comment (soumik12345) [inline sagan/models/sagan.py]: `FLAGS.experiment_configs.latent_dim` and `FLAGS.experiment_configs.gp_weight` aren't valid literal names in python. It should be 

```python3
def __init__(self, configs: ml_collections.ConfigDicts, **kwargs):
    self.latent_dim = configs.latent_dim
    self.gp_weight = configs.gp_weight
    self.generator = build_generator(configs.image_size, configs.latent_dim)
    self.discriminator = build_discriminator(configs.image_size)
```

File: sagan/models/sagan.py [hunk 22]
```diff
@@ -0,0 +1,104 @@
+from absl import app
+from absl import flags
+
+from ml_collections.config_flags import config_flags
```
Comment (soumik12345) [inline sagan/models/sagan.py]: Seems like an unnecessary import

File: sagan/utils.py [hunk 23]
```diff
@@ -0,0 +1,21 @@
+from absl import app
+from absl import flags
+
+from ml_collections.config_flags import config_flags
```
Comment (soumik12345) [inline sagan/utils.py:4]: Another unnecessary import

File: train.py [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+## Compile the model
+
+### Define the optimizers
+generator_optimizer = keras.optimizers.Adam(learning_rate=0.0001, beta_1=0, beta_2=0.9)
+discriminator_optimizer = keras.optimizers.Adam(
+    learning_rate=0.0004, beta_1=0, beta_2=0.9
+)
+
+### Define the loss function
+loss_fn = tf.keras.losses.BinaryCrossentropy(from_logits=True, label_smoothing=0.1)
```
Comment (soumik12345) [inline train.py]: This is not needed, we are using Wasserstein loss

File: train.py [hunk 25]
```diff
@@ -0,0 +1,54 @@
...
+
+sagan = sagan.compile(discriminator_optimizer, generator_optimizer, loss_fn)
```
Comment (soumik12345) [inline train.py]: No need to pass `loss_fn`

File: train.py [hunk 26]
```diff
@@ -0,0 +1,54 @@
...
+
+## Train the model
+for epoch in range(FLAGS.experiment_configs.epochs):
```
Comment (soumik12345) [inline train.py]: Calling `model.fit()` inside for loop is not a good practice.

**Review by soumik12345 — CHANGES_REQUESTED** [review]
@sayannath Feedback round 3

File: sagan/models/sagan.py [hunk 27]
```diff
@@ -0,0 +1,104 @@
...
+FLAGS = flags.FLAGS
+config_flags.DEFINE_config_file("experiment_configs")
+
+class SelfAttentionGAN(keras.Model):
+    def __init__(self, configs: ml_collections.ConfigDicts, **kwargs):
+        super(SelfAttentionGAN, self).__init__(**kwargs)
+        self.latent_dim = configs.latent_dim
+        self.gp_weight = configs.gp_weight
+        self.generator = build_generator(configs.image_size, configs.latent_dim)
+        self.discriminator = build_discriminator(FLAGS.experiment_configs.image_size)
```
Comment (soumik12345) [inline sagan/models/sagan.py:22]:
```python3
self.discriminator = build_discriminator(configs.image_size)
```

File: train.py [hunk 28]
```diff
@@ -0,0 +1,45 @@
...
+
+## Define the hyperparameters
+generator_save_path = configs.save_generator_model_path
+discriminator_save_path = configs.save_discriminator_model_path
+
+generator = build_generator()
+discriminator = build_discriminator()
+dataset = DataLoader.get_dataset()
+
+sagan = SelfAttentionGAN(discriminator, generator, configs.latent_dim) # Will come from the config file
```
Comment (soumik12345) [inline train.py]:
```python3
sagan = SelfAttentionGAN(discriminator, generator, FLAGS.experiment_configs.latent_dim) 
```

File: train.py [hunk 29]
```diff
@@ -0,0 +1,45 @@
...
+
+## Compile the model
+
+### Define the optimizers
+generator_optimizer = keras.optimizers.Adam(learning_rate=0.0001, beta_1=0, beta_2=0.9)
+discriminator_optimizer = keras.optimizers.Adam(
+    learning_rate=0.0004, beta_1=0, beta_2=0.9
+)
```
Comment (soumik12345) [inline train.py:29]:
```python3
generator_optimizer = keras.optimizers.Adam(learning_rate=FLAGS.experiment_configs.generator_learning_rate, beta_1=FLAGS.experiment_configs.beta_one, beta_2=FLAGS.experiment_configs.beta_two)
discriminator_optimizer = keras.optimizers.Adam(
    learning_rate=FLAGS.experiment_configs.discriminator_learning_rate, beta_1=FLAGS.experiment_configs.beta_one, beta_2=FLAGS.experiment_configs.beta_two
)
```

File: train.py [hunk 30]
```diff
@@ -0,0 +1,45 @@
...
+
+## Compile the Self-Attention GAN model
+sagan = sagan.compile(discriminator_optimizer, generator_optimizer)
+
+## Setting up Callback
+cbk = GANMonitor(num_img=3, latent_dim=FLAGS.experiment_configs.lentent_dim)
```
Comment (soumik12345) [inline train.py]: No need for this, we would integrate with W&B anyway

File: train.py [hunk 31]
```diff
@@ -0,0 +1,45 @@
...
+
+## Train the model
+sagan.fit(dataset, batch_size=FLAGS.experiment_configs.batch_size, epochs=FLAGS.experiment_configs.epochs, callbacks=[cbk])
```
Comment (soumik12345) [inline train.py]: `ModelCheckpoint` callback is a necessary one here.

**Review by soumik12345 — CHANGES_REQUESTED** [review]
@sayannath Feedback Round 4

**Review by soumik12345 — APPROVED** [review]
@sayannath Updates look fine. Please fix the pipeline, we can merge then.
